### ISSUE TITLE
I'm going to pretrain Albert on my own data. What am I going to do?What format is the training data?

### ISSUE THREAD

**Issue by bigzhouj** [google-research/ALBERT#165]

**Comment by arrrrrmin** [comment]
Hey @bigzhouj I'm performing the same task. Here is a little checklist you can go though:

* train a tokenizer on your data (it'll perform word/subword splitting and tokenization)

`
spm_train \
--input ../all.txt --model_prefix=30k-clean --vocab_size=30000 --logtostderr
--pad_id=0 --unk_id=1 --eos_id=-1 --bos_id=-1
--control_symbols=[CLS],[SEP],[MASK]
--user_defined_symbols=(,),\”,-,.,–,£,€
--shuffle_input_sentence=true --input_sentence_size=10000000
--character_coverage=0.99995 --model_type=unigram
`
The way in which SentencePiece is expecting the data is documented [here](https://github.com/google/sentencepiece/blob/master/python/README.md).

* Build training data, using the tokenizer and write .tfrecord files. For more info on these see [tfrecord-documentation](https://www.tensorflow.org/tutorials/load_data/tfrecord)
`
time python3 create_pretraining_data.py \
  --input_file="../all.txt.1"  \
  --output_file="../examples.tfrecord" \
  --vocab_file="../tokenizer.vocab" \
  --spm_model_file="../tokenizer.model"
`

The rest should be documented in ALBERT/README.md. I it hope this helps a bit.

**Comment by arrrrrmin** [comment]
Ah sorry, forgot to mention [this](https://github.com/google-research/ALBERT/blob/4c601a1c096e85b1639460ea6b890ba5a8d2b3ee/create_pretraining_data.py#L218):

> Input file format:
(1) One sentence per line. These should ideally be actual sentences, not
entire paragraphs or arbitrary spans of text. (Because we use the
sentence boundaries for the "next sentence prediction" task).
(2) Blank lines between documents. Document boundaries are needed so
that the "next sentence prediction" task doesn't span between documents.

Should also be helpful for your data prep.

**Comment by 008karan** [comment]
@arrrrrmin I think you can help here
I have generated pretraining data using [https://github.com/kamalkraj/ALBERT-TF2.0](url)
because this supports training with multi GPU. I am doing this for the Hindi language with 22gb of data. Generating pretraining data itself took 1 month!
So I have `meta_data` file associated with each tf.record file. I have added all the `train_data_size` values from all the `meta_data` files to make one `meta_data` file because in `run_pretraining.py` requires it.  So my final `meta_data` file which looks something like this:
```
{
    "task_type": "albert_pretraining",
    "train_data_size": 596972848,
    "max_seq_length": 512,
    "max_predictions_per_seq": 20
}
```
Here number of training steps are calculated as below:
`num_train_steps = int(total_train_examples / train_batch_size) * num_train_epochs`
So `total_train_examples` is 596972848 hence I am getting  `num_train_steps` to be 9327700 with batch size of 64 and with 1 epoch only. I saw that in readme here `num_train_steps=125000`. I am not getting whats went wrong here. 

With such huge train steps, it will take forever to train Albert. Even if I make batch size to 512 with 1 epoch only the training step will be 1165962 which is still huge!
As Albert was trained on very huge data why there are only 125000 steps only?
Want to know-how many epochs are there in Albert training for English?

Can you suggest what should I do now?

**Comment by AminTaheri23** [comment]
> Generating pretraining data itself took 1 month!

@008karan 

What hardware did you use to train? Can you specify, please?

**Comment by 008karan** [comment]
i was using 4 V100 16 gb.
